oak/server/rust/oak_runtime/src/node/http/mod.rs

```diff
@@ -0,0 +1,520 @@
+        .basic_scheduler()
+        // Enables the I/O driver.
+        // Necessary for using net, process, signal, and I/O types on the Tokio runtime.
+        .enable_io()
+        // Enables the time driver.
+        // Necessary for creating a Tokio Runtime.
+        .enable_time()
+        .build()
+        .expect("Couldn't create Async runtime")
+}
+
+/// [`HttpRequestHandler`] handles HTTP requests from a client and sends HTTP responses back.
+#[derive(Clone)]
+struct HttpRequestHandler {
+    /// Reference to the Runtime in the context of this HTTP server pseudo-Node.
+    runtime: RuntimeProxy,
+    /// Channel handle used for writing HTTP invocations.
+    invocation_channel: oak_abi::Handle,
+}
+
+impl HttpRequestHandler {
+    fn handle(
+        &self,
+        request: HttpRequest,
+    ) -> Pin<Box<dyn Future<Output = Result<Response<Body>, OakStatus>> + Send + Sync>> {
+        let handler = self.clone();
+        let future = async move {
+            let oak_label = get_oak_label(&request)?;
+            info!(
+                "handling HTTP/2 request; request size: {} bytes, label: {:?}",
+                request.body.len(),
+                oak_label
+            );
+
+            debug!("inject the request into the Oak Node");
+            let response = handler
+                .inject_http_request(request, &oak_label)
+                .map_err(|_| OakStatus::ErrInternal)?;
+
+            Ok(response.to_response())
+        };
+
+        Box::pin(future)
+    }
+
+    fn inject_http_request(
+        &self,
+        request: HttpRequest,
+        label: &Label,
+    ) -> Result<HttpResponseIterator, ()> {
+        // Create a pair of temporary channels to pass the HTTP/2 request and to receive the
+        // response.
+        let pipe = Pipe::new(&self.runtime.clone(), label)?;
+
+        // Serialize HTTP/2 request into a message.
+        let mut message = crate::NodeMessage {
+            data: vec![],
+            handles: vec![],
+        };
+
+        request.encode(&mut message.data).map_err(|err| {
+            error!("Couldn't serialize HttpRequest message: {}", err);
+        })?;
+
+        // Put the HTTP request message inside the per-invocation request channel.
+        pipe.insert_message(&self.runtime, message)?;
+
+        // Send an invocation message (with attached handles) to the Oak Node.
+        pipe.send_invocation(&self.runtime, self.invocation_channel)?;
+
+        // Close all local handles except for the one that allows reading responses.
+        pipe.close(&self.runtime);
+
+        Ok(HttpResponseIterator {
+            runtime: self.runtime.clone(),
+            response_reader: pipe.response_reader,
+        })
+    }
+}
+
+// A pair of temporary channels to pass the HTTP request and to receive the response.
+struct Pipe {
+    request_writer: oak_abi::Handle,
+    request_reader: oak_abi::Handle,
+    response_writer: oak_abi::Handle,
+    response_reader: oak_abi::Handle,
```

Comment:
Could we use more expressive `Sender` and `Receiver` types here instead?

https://github.com/project-oak/oak/blob/main/oak/server/rust/oak_runtime/src/io.rs